login/handlers.go

```diff
@@ -76,34 +77,89 @@ func (server *Server) handleLoginRequest(conn mnet.Client, reader mpacket.Reader
 	// Already online = 7, System error = 9, Too many requests = 10, Older than 20 = 11, Master cannot login on this IP = 13
 
 	if result <= 0x01 {
-		conn.SetLogedIn(true)
 		conn.SetGender(gender)
 		conn.SetAdminLevel(adminLevel)
 		conn.SetAccountID(accountID)
+	}
 
-		_, err := common.DB.Exec("UPDATE accounts set isLogedIn=1 WHERE accountID=?", accountID)
+	conn.Send(packetLoginResponse(result, accountID, gender, adminLevel > 0, username, isBanned))
+}
 
-		if err != nil {
-			log.Println("Database error with approving login of accountID", accountID, err)
-		} else {
-			log.Println("User", accountID, "has logged in from", conn)
-		}
+func (server *Server) handlePinRegistration(conn mnet.Client, reader mpacket.Reader) {
+	b1 := reader.ReadByte()
+
+	if b1 == 0 { // Client canceled pin change request
+		conn.Send(packetCancelPin())
+		return
+	}
+	reader.Skip(2)
+
+	accountID := conn.GetAccountID()
+	pin := string(reader.GetRestAsBytes())
+
+	_, err := common.DB.Exec("UPDATE accounts SET pin=? WHERE accountID=?", pin, accountID)
+	if err != nil {
+		log.Println("handlePinRegistration database pin update issue for accountID:", accountID, err)
 	}
 
-	conn.Send(packetLoginResponce(result, accountID, gender, adminLevel > 0, username, isBanned))
+	conn.Send(packetRequestPin())
+
 }
 
 func (server *Server) handleGoodLogin(conn mnet.Client, reader mpacket.Reader) {
 	server.migrating[conn] = false
-	var username, password string
-
 	accountID := conn.GetAccountID()
 
-	err := common.DB.QueryRow("SELECT username, password FROM accounts WHERE accountID=?", accountID).
-		Scan(&username, &password)
+	if server.withPin {
+		var pinDB string
+		err := common.DB.QueryRow("SELECT pin FROM accounts WHERE accountID=?", accountID).
+			Scan(&pinDB)
+
+		if err != nil {
+			log.Println("handleCheckLogin database retrieval issue for accountID:", accountID, err)
+		}
+
+		b1 := reader.ReadByte()
+		b2 := reader.ReadByte()
+
+		if b1 == 1 && b2 == 1 { // First attempt, request for pin
+			if len(pinDB) == 0 {
+				conn.Send(packetRegisterPin())
+			} else {
+				conn.Send(packetRequestPin())
+			}
+			return
```

Comment:
I'm not a fan of having lots of return statements like this. See if the logic can be done in a way where there aren't as many return statements.

Reply:
done